tests/test_histogram.py

```diff
@@ -83,30 +83,79 @@ def test_compare_histograms_optimized_vs_general():
+                                          binned_feature, ordered_gradients)
+    hist_ghc = _build_histogram(n_bins, sample_indices, binned_feature,
+                                ordered_gradients, ordered_hessians)
+
+    hist_naive = _build_histogram_naive(n_bins, sample_indices, binned_feature,
+                                        ordered_gradients, ordered_hessians)
+
+    for hist in (hist_gc_root, hist_ghc_root, hist_gc, hist_gc, hist_ghc):
+        assert_array_equal(hist['count'], hist_naive['count'])
+        assert_allclose(hist['sum_gradients'], hist_naive['sum_gradients'])
+    for hist in (hist_ghc_root, hist_ghc):
+        assert_allclose(hist['sum_hessians'], hist_naive['sum_hessians'])
+    for hist in (hist_gc_root, hist_gc):
+        assert_array_equal(hist['sum_hessians'], np.zeros(n_bins))
+
+
+@pytest.mark.parametrize("constant_hessian", [True, False])
+def test_hist_subtraction(constant_hessian):
+    # Make sure the histogram subtraction trick gives the same result as the
+    # classical method.
+    rng = np.random.RandomState(42)
+    n_samples = 10
+    n_bins = 5
+    sample_indices = np.arange(n_samples).astype(np.uint32)
+    binned_feature = rng.randint(0, n_bins - 1, size=n_samples, dtype=np.uint8)
+    ordered_gradients = rng.randn(n_samples).astype(np.float32)
+    if constant_hessian:
+        ordered_hessians = np.ones(n_samples, dtype=np.float32)
+    else:
+        ordered_hessians = rng.lognormal(size=n_samples).astype(np.float32)
+
+    if constant_hessian:
+        hist_parent = _build_histogram_no_hessian(n_bins, sample_indices,
+                                                  binned_feature,
+                                                  ordered_gradients)
+    else:
+        hist_parent = _build_histogram(n_bins, sample_indices, binned_feature,
+                                       ordered_gradients, ordered_hessians)
+
+    mask = rng.randint(0, 2, n_samples).astype(np.bool)
+
+    sample_indices_left = sample_indices[mask]
+    ordered_gradients_left = ordered_gradients[mask]
+    ordered_hessians_left = ordered_hessians[mask]
+    if constant_hessian:
+        hist_left = _build_histogram_no_hessian(n_bins, sample_indices_left,
+                                                binned_feature,
+                                                ordered_gradients_left)
+    else:
+        hist_left = _build_histogram(n_bins, sample_indices_left,
+                                     binned_feature, ordered_gradients_left,
+                                     ordered_hessians_left)
+
+    sample_indices_right = sample_indices[~mask]
+    ordered_gradients_right = ordered_gradients[~mask]
+    ordered_hessians_right = ordered_hessians[~mask]
+    if constant_hessian:
+        hist_right = _build_histogram_no_hessian(n_bins, sample_indices_right,
+                                                 binned_feature,
+                                                 ordered_gradients_right)
```

Comment:
same here: `hist_right_2 = _build_histogram(..., ordered_hessians_right)` and check that `hist_right` is all close to `hist_right_2`.